### PR TITLE
ENH: Added ``kulczynski1`` to ``scipy.spatial.distance``

### LINKED ISSUE #2009: "Kulsinski" dissimilarity seems wrong (Trac #1484)
_Original ticket http://projects.scipy.org/scipy/ticket/1484 on 2011-07-28 by trac user muellner, assigned to trac user peridot._

The "Kulsinski" dissimilarity in the module scipy.spatial.distance seems wrong.

First, the function does not do what the documentation 

http://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.kulsinski.html#scipy.spatial.distance.kulsinski

says: it always returns integer values. Maybe this is a forgotten float(...) in scipy/spatial/distance.py, line 421.

Second, the definition does not agree what I find elsewhere on the net. Compare

http://genome.jouy.inra.fr/doc/genome/statistiques/R-2.6.0/library/prabclus/html/kulczynski.html

and

http://sysbio.oxfordjournals.org/content/55/1/170.full

Can someone please check whether the SciPy definition makes sense and/or is intenionally different from the definition found in R?

### PR DESCRIPTION
**Pull request by czgdp1807**
#### Reference issue

Closes https://github.com/scipy/scipy/issues/2009
Point 5 in https://github.com/scipy/scipy/pull/3163#issuecomment-34224341
https://github.com/scipy/scipy/issues/2009#issuecomment-898674427

#### Additional information

cc @jaimefrio @rgommers

### CODE REVIEW

**Comment by tupui** [comment]
Just a side note, if it's WIP, please use the draft functionality of GitHub instead of putting it in the name. Better for filtering etc.

**Comment by czgdp1807** [comment]
I can't see an option now to convert this to Draft. I will keep this in mind in the next PR.

**Comment by tupui** [comment]
> I can't see an option now to convert this to Draft. I will keep this in mind in the next PR.

It was in the reviewer section. I did it for you.

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -934,6 +934,47 @@ def kulsinski(u, v, w=None):
...
+        Input array.
+    v : (N,) array_like, bool
+        Input array.
+    w : (N,) array_like, optional
+        The weights for each value in `u` and `v`. Default is None,
+        which gives each value a weight of 1.0
+
+    Returns
+    -------
+    kulczynski1 : double
```
Comment (tupui) [inline scipy/spatial/distance.py:964]: We don't have double in Python, should be float here.
```suggestion
    kulczynski1 : float
```

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -934,6 +934,47 @@ def kulsinski(u, v, w=None):
...
+         \\frac{c_{TT}}
+              {c_{FT} + c_{TF}}
+
+    where :math:`c_{ij}` is the number of occurrences of
+    :math:`\\mathtt{u[k]} = i` and :math:`\\mathtt{v[k]} = j` for
+    :math:`k < n`.
+
+    Parameters
+    ----------
+    u : (N,) array_like, bool
```
Comment (tupui) [inline scipy/spatial/distance.py:954]: Can you add an example with using `bool`? Is this vectorized?

Comment (czgdp1807) [reply scipy/spatial/distance.py:954]: `_nbool_correspond_all` is vectorized. So, seems like yes, this function is vectorized as well. I have followed pattern similar to the other functions in this file.

**Comment by czgdp1807** [comment]
I have added documentation. It seems like there is nothing more to be done for reviews. So, marking it as `Ready For Review`.

File: scipy/spatial/distance.py [hunk 3]
```diff
@@ -931,6 +933,57 @@ def kulsinski(u, v, w=None):
...
+    -------
+    kulczynski1 : float
+        The Kulczynski 1 distance between vectors `u` and `v`.
+
+    Examples
+    --------
+    >>> from scipy.spatial import distance
+    >>> distance.kulczynski1([1, 0, 0], [0, 1, 0])
+    0.0
+    >>> distance.kulczynski1([True, False, False], [True, True, False])
```
Comment (tupui) [inline scipy/spatial/distance.py:972]: Thanks. My question about vectorization is does this works `distance.kulczynski1([True, False, False], [True])`? And if not, do we have a comprehensive warning for it?

File: scipy/spatial/distance.py [hunk 4]
```diff
@@ -931,6 +933,57 @@ def kulsinski(u, v, w=None):
     return (ntf + nft - ntt + n) / (ntf + nft + n)
 
 
+def kulczynski1(u, v, w=None):
```
Comment (tupui) [inline scipy/spatial/distance.py:936]: Make extra parameters keywords only.
```suggestion
def kulczynski1(u, v, *, w=None):
```

File: scipy/spatial/distance.py [hunk 5]
```diff
@@ -931,6 +933,57 @@ def kulsinski(u, v, w=None):
...
+    """
+    Compute the Kulczynski 1 dissimilarity between two boolean 1-D arrays.
+
+    The Kulczynski 1 dissimilarity between two boolean 1-D arrays `u` and `v`,
+    is defined as
+
+    .. math::
+
+         \\frac{c_{TT}}
+              {c_{FT} + c_{TF}}
```
Comment (tupui) [inline scipy/spatial/distance.py:946]: What is TT, FT, TF? It does not relate to `c_{ij}`

File: scipy/spatial/distance.py [hunk 6]
```diff
@@ -931,6 +933,57 @@ def kulsinski(u, v, w=None):
...
+
+    where :math:`c_{ij}` is the number of occurrences of
+    :math:`\\mathtt{u[k]} = i` and :math:`\\mathtt{v[k]} = j` for
+    :math:`k < n`.
```
Comment (tupui) [inline scipy/spatial/distance.py:950]: What are k and n?

File: scipy/spatial/distance.py [hunk 7]
```diff
@@ -931,6 +933,57 @@ def kulsinski(u, v, w=None):
     return (ntf + nft - ntt + n) / (ntf + nft + n)
 
 
+def kulczynski1(u, v, w=None):
+    """
+    Compute the Kulczynski 1 dissimilarity between two boolean 1-D arrays.
```
Comment (tupui) [inline scipy/spatial/distance.py:938]: Is the value bounded? How to interpret a low/high value?

File: scipy/spatial/distance.py [hunk 8]
```diff
@@ -931,6 +933,57 @@ def kulsinski(u, v, w=None):
...
+    w : (N,) array_like, optional
+        The weights for each value in `u` and `v`. Default is None,
+        which gives each value a weight of 1.0
+
+    Returns
+    -------
+    kulczynski1 : float
+        The Kulczynski 1 distance between vectors `u` and `v`.
+
+    Examples
```
Comment (tupui) [inline scipy/spatial/distance.py:967]: Version added and see also. But is `kulsinski` related to `kulczynski1`? Is there a typo somewhere? If related, maybe add a note about the difference.

```suggestion
        The Kulczynski 1 distance between vectors `u` and `v`.

    .. versionadded:: 1.8.0

    See Also
    --------
    kulsinski

    Examples
```

File: scipy/spatial/distance.pyi [hunk 9]
```diff
@@ -153,6 +153,10 @@ def kulsinski(
     u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
 ) -> np.float64: ...
 
+def kulczynski1(
+    u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
```
Comment (tupui) [inline scipy/spatial/distance.pyi:157]: But also bool no?

**Review by tupui — CHANGES_REQUESTED** [review]
Great start! I have a few more comments. I did not yet look at the C (and someone else could be more suited).

File: scipy/spatial/distance.py (hunk 3 above)
Comment (czgdp1807) [reply scipy/spatial/distance.py:972]: Yes. It works.

File: scipy/spatial/distance.py (hunk 4 above)
Comment (czgdp1807) [reply scipy/spatial/distance.py:936]: Other functions like, `hamming`, `jaccard`, `kulsinski`, etc, follow (u, v, w=None) signature in this file. Should I also make them, `(u, v, *, w=None)`.

File: scipy/spatial/distance.py (hunk 5 above)
Comment (czgdp1807) [reply scipy/spatial/distance.py:946]: The documentation is also following the same pattern, `hamming`, `jaccard`, etc. Should we make the change there too?

File: scipy/spatial/distance.py (hunk 6 above)
Comment (czgdp1807) [reply scipy/spatial/distance.py:950]: Same as https://github.com/scipy/scipy/pull/14588#discussion_r691793219

File: scipy/spatial/distance.pyi (hunk 9 above)
Comment (czgdp1807) [reply scipy/spatial/distance.pyi:157]: This is following the same pattern as other functions. However, I checked and found that other functions also accept `bool` and not only `ArrayLike`. So, should we make the change at other places too? If yes, then in this PR or in a follow up?

File: scipy/spatial/distance.py (hunk 4 above)
Comment (tupui) [reply scipy/spatial/distance.py:936]: We cannot change previous functions without breaking backward compatibility here. So just do it for new functions or when adding new parameters.

File: scipy/spatial/distance.py (hunk 5 above)
Comment (tupui) [reply scipy/spatial/distance.py:946]: Let's get this first right here and also contained. You/we can then make a followup which addresses just the documentation of similar functions. Also Hamming seems to be better as does not introduce more notations (I am not saying this is the best OC. It's still missing a definition for `k` for instance).

File: scipy/spatial/distance.py (hunk 3 above)
Comment (tupui) [reply scipy/spatial/distance.py:984]: Then IMO an example would be nice.

File: scipy/spatial/distance.pyi (hunk 9 above)
Comment (tupui) [reply scipy/spatial/distance.pyi:157]: Actually this is ok as `ArrayLike` take care of `bool` as well.

**Review by tupui — COMMENTED** [review]
Side note: when replying in discussions, try to go in the file tab an make a review instead. This allow to group your replies and only send a single email/notification 😉

File: scipy/spatial/distance.py (hunk 8 above)
Comment (tupui) [reply scipy/spatial/distance.py]: Ok forget about the see also section. I just read the issues and no we do not want to reference a deprecated thing.

**Comment by tupui** [comment]
I am not sure about the name here. If `kulczynski1` is meant to replace `kulsinski` then I would prefer `kulczynski`. Or are we going to add `kulczynski2` or something else?

File: scipy/spatial/distance.py (hunk 8 above)
Comment (tupui) [reply scipy/spatial/distance.py]: Can you also add a reference for the metric?

**Comment by czgdp1807** [comment]
> If `kulczynski1` is meant to replace `kulsinski` then I would prefer `kulczynski`.

There are two flavours of Kulczynski, `#1` and `#2`. If we use only, `kulczynski` as the name then it won't reflect which one we are refering to until we read the documentation. `kulczynski1` does that (makes code reading easier). IMHO, it would be better to use `kulczynski1` as it clearly reflects the meaning even if we do not add `kulczynski2` in future.

**Comment by tupuinui** [comment]
> > If `kulczynski1` is meant to replace `kulsinski` then I would prefer `kulczynski`.
> 
> There are two flavours of Kulczynski, `#1` and `#2`. If we use only, `kulczynski` as the name then it won't reflect which one we are refering to until we read the documentation. `kulczynski1` does that (makes code reading easier). IMHO, it would be better to use `kulczynski1` as it clearly reflects the meaning even if we do not add `kulczynski2` in future.

Why not, but I just would like to avoid us adding `kulczynski2` later. This would, IMO, not be pretty and could be handled with a parameter (except OC if these would need very different inputs/outputs)

**Comment by tupui** [comment]
> > If `kulczynski1` is meant to replace `kulsinski` then I would prefer `kulczynski`.
> 
> There are two flavours of Kulczynski, `#1` and `#2`. If we use only, `kulczynski` as the name then it won't reflect which one we are refering to until we read the documentation. `kulczynski1` does that (makes code reading easier). IMHO, it would be better to use `kulczynski1` as it clearly reflects the meaning even if we do not add `kulczynski2` in future.

Why not, but I just would like to avoid us adding kulczynski2 later. This would, IMO, not be pretty and could be handled with a parameter (except OC if these would need very different inputs/outputs)

**Comment by czgdp1807** [comment]
> This would, IMO, not be pretty and could be handled with a parameter

You mean, `kulczynski(u, v, *, w=None, type=1)`? I thought of something similar initially but later on went with `kulczynski1` to avoid an extra odd looking `type` parameter when all the other APIs in the same file don't have it. I agree, `kulczynski1` isn't good looking.

**Comment by tupui** [comment]
> > This would, IMO, not be pretty and could be handled with a parameter
> 
> You mean, `kulczynski(u, v, *, w=None, type=1)`? I thought of something similar initially but later on went with `kulczynski1` to avoid an extra odd looking `type` parameter when all the other APIs in the same file don't have it. I agree, `kulczynski1` isn't good looking.

I mean if you also don't like `kulczynski1`, then I would say we go for `kulczynski`. But you can leave it like that and wait for others' opinions. (It would not be type as it's a reserved keyword but maybe `dimension`, `method` or whatever else it represents).

**Comment by czgdp1807** [comment]
I have addressed the reviews.

**Review by tupui — COMMENTED** [review]
@czgdp1807 it's in good shape! You're still missing some doc (cf open discussions).

**Review by tupui — CHANGES_REQUESTED** [review]

**Comment by czgdp1807** [comment]
I have addressed the reviews. The windows tests were cancelled due to their long duration, rest all pass.

File: scipy/spatial/distance.py [hunk 10]
```diff
@@ -931,6 +933,79 @@ def kulsinski(u, v, w=None):
...
+
+    .. versionadded:: 1.8.0
+
+    References
+    ----------
+    .. [1] Kulczynski S. Zespoly roslin w Pieninach. Bulletin
+           International de lAcademie Polonaise des Sciences
+           et des Lettres, Classe des Sciences Mathe matiques
+           et Naturelles, Se rie B (Sciences Naturelles). 1927;
+           Supplement II: 57-203.
```
Comment (tupui) [inline scipy/spatial/distance.py:985]:
```suggestion
    .. [1] Kulczynski S. et al. Bulletin
           International de l'Academie Polonaise des Sciences
           et des Lettres, Classe des Sciences Mathematiques
           et Naturelles, Serie B (Sciences Naturelles). 1927;
           Supplement II: 57-203.
```

**Review by tupui — APPROVED** [review]
Thanks @czgdp1807, let's get this in. If we want to uniformize the doc, it can be done latter.

**Comment by czgdp1807** [comment]
Thanks for the help and pin-pointing several small but important things @tupui.

**Pull request merged**